app/src/androidTest/java/br/com/simplepass/loadingbutton/CircularProgressButtonTest.kt

```diff
@@ -0,0 +1,108 @@
+import androidx.test.espresso.matcher.ViewMatchers.isDisplayed
+import androidx.test.espresso.matcher.ViewMatchers.isEnabled
+import androidx.test.espresso.matcher.ViewMatchers.withText
+import androidx.test.espresso.matcher.ViewMatchers.withEffectiveVisibility
+import androidx.test.platform.app.InstrumentationRegistry
+import androidx.test.rule.ActivityTestRule
+import androidx.test.runner.AndroidJUnit4
+import br.com.simplepass.loadingbutton.customViews.CircularProgressButton
+import br.com.simplepass.loadingbuttonsample.MainActivity
+import br.com.simplepass.loadingbuttonsample.R
+import org.hamcrest.Matchers.containsString
+import org.hamcrest.Matchers.not
+import org.hamcrest.core.AllOf.allOf
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+
+/**
+ * [Testing Fundamentals](http://d.android.com/tools/testing/testing_android.html)
+ */
+@RunWith(AndroidJUnit4::class)
+class CircularProgressButtonTest {
+
+    @JvmField
+    @Rule
+    var activityTestRule = ActivityTestRule(MainActivity::class.java, true, false)
+
+    private lateinit var app: Application
+
+    @Before
+    fun setup() {
+        val instrumentation = InstrumentationRegistry.getInstrumentation()
+        app = instrumentation.targetContext.applicationContext as Application
+    }
+
+    @Test
+    fun testTextVisibleBeforeSpinnerAnimation() {
+        val mainActivity = activityTestRule.launchActivity(Intent(app, MainActivity::class.java))
+
+        onView(withId(R.id.buttonTest1))
+                .check(matches(allOf<View>(
+                        isDisplayed(), withEffectiveVisibility(ViewMatchers.Visibility.VISIBLE), isEnabled())))
+
+        onView(withId(R.id.buttonTest1))
+                .check(matches(withText(
+                        containsString(mainActivity.resources.getString(R.string.send)))))
+
+        onView(withId(R.id.buttonTest1)).perform(click())
+
+        onView(withId(R.id.buttonTest1))
+                .check(matches(not<View>(withText(
+                        containsString(mainActivity.resources.getString(R.string.send))))))
+    }
+
+    @Test
+    fun testTextVisibleBeforeAndAfterSpinnerAnimation() {
+        val mainActivity = activityTestRule.launchActivity(Intent(app, MainActivity::class.java))
+
+        onView(withId(R.id.buttonTest2))
+                .check(matches(allOf<View>(
+                        isDisplayed(), withEffectiveVisibility(ViewMatchers.Visibility.VISIBLE), isEnabled())))
+
+        onView(withId(R.id.buttonTest2))
+                .check(matches(withText(
+                        containsString(mainActivity.resources.getString(R.string.send)))))
+
+        onView(withId(R.id.buttonTest2)).perform(click())
+
+        val circularProgressButton = mainActivity.findViewById<CircularProgressButton>(R.id.buttonTest2)
+        Espresso.registerIdlingResources(ViewVisibilityIdlingResource(circularProgressButton, VISIBLE))
```

Comment:
This idling resource is never waiting for anything, since the button is never going out of sight. Did you mean to wait for the text inside the button?
if that's the case, I am not sure the `ViewVisibilityIdlingResource` is going to work... the text inside the button is not a View, after all. We need to thing about a solution here

Reply:
I think this should work:

```
class TextIdlingResource(private val view: TextView, private val expectedText: String) : IdlingResource {

    private var idle: Boolean = false
    private var resourceCallback: ResourceCallback? = null

    init {
        this.idle = false
        this.resourceCallback = null
    }

    override fun getName(): String {
        return TextIdlingResource::class.java.simpleName
    }

    override fun isIdleNow(): Boolean {
        idle = idle || view.text.toString() == expectedText

        if (idle) {
            if (resourceCallback != null) {
                resourceCallback!!.onTransitionToIdle()
            }
        }

        return idle
    }

    override fun registerIdleTransitionCallback(resourceCallback: ResourceCallback) {
        this.resourceCallback = resourceCallback
    }
}
```

Reply:
I tried using this, but I'm not sure the implementation works as expected. Unless the value of `View#getText` is changing here, this `IdlingResource` will immediately transition to idle for a matching view. The value of `getText` does not change along with view visibility.

Reply:
I tested and it worked as I expected '.' 
The text change during the animation, it is set to `null`, otherwise it wouldn't disappear. So the view.text should return null. But, now that I am thing about this Idling resource... if we add a possibility to keep the text as the animation goes on, this idling resource won't make sense anymore... We should use the button state instead something that might change the logic. 

This version of the Idling Resource is better:

```
class ProgressButtonStateIdlingResource(
    private val progressButton: ProgressButton,
    private val expectedState: State = State.IDLE
) : IdlingResource {

    private var idle: Boolean = false
    private var resourceCallback: ResourceCallback? = null

    init {
        this.idle = false
        this.resourceCallback = null
    }

    override fun getName(): String {
        return progressButton::class.java.simpleName
    }

    override fun isIdleNow(): Boolean {
        idle = idle || progressButton.getState() == expectedState

        if (idle) {
            if (resourceCallback != null) {
                resourceCallback!!.onTransitionToIdle()
            }
        }

        return idle
    }

    override fun registerIdleTransitionCallback(resourceCallback: ResourceCallback) {
        this.resourceCallback = resourceCallback
    }
}
```

Reply:
Right now, the getState() is not available. I'll expose this method.

Reply:
https://github.com/leandroBorgesFerreira/LoadingButtonAndroid/pull/103

Reply:
+1 -- something keyed off the state sounds like the most robust solution for this specific use case. That said, the build is now passing, including the tests as they are currently written. I see a few paths forward from here:

1. Review the PR again in it's current form, and proceed with no structural changes if approved.
2. Update this PR to remove `Wait` and replace it with `ProgressButtonStateIdlingResource`.
3. Move forward with the PR in it's current form and file an issue to capture the refactor to `ProgressButtonStateIdlingResource`.

Thoughts?